helix-core/src/main/java/org/apache/helix/manager/zk/ZkBucketDataAccessor.java

```diff
@@ -202,126 +218,156 @@ public void disconnect() {
+
+    // Combine buckets into one byte array
+    int copyPtr = 0;
+    for (int i = 0; i < numBuckets; i++) {
+      if (i == numBuckets - 1) {
+        // Special treatment for the last bucket
+        System.arraycopy(buckets.get(i), 0, compressedRecord, copyPtr, dataSize % bucketSize);
+      } else {
+        System.arraycopy(buckets.get(i), 0, compressedRecord, copyPtr, bucketSize);
+        copyPtr += bucketSize;
       }
     }
-    throw new HelixException(String.format("Could not acquire lock for read. Path: %s", path));
+
+    // Decompress the byte array
+    ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(compressedRecord);
+    byte[] serializedRecord;
+    try {
+      serializedRecord = GZipCompressionUtil.uncompress(byteArrayInputStream);
+    } catch (IOException e) {
+      throw new HelixException(String.format("Failed to decompress path: %s!", path), e);
+    }
+
+    // Deserialize the record to retrieve the original
+    ZNRecord originalRecord = (ZNRecord) _zkSerializer.deserialize(serializedRecord);
+    return new HelixProperty(originalRecord);
+  }
+
+  @Override
+  public void close() {
+    disconnect();
   }
 
   /**
-   * Returns the last string element in a split String array by /.
+   * Deletes all versions that are older than the currentVersion given.
    * @param path
-   * @return
+   * @param currentVersion
    */
-  private String extractIdFromPath(String path) {
-    String[] splitPath = path.split("/");
-    return splitPath[splitPath.length - 1];
+  private void deleteStaleVersions(String path, long currentVersion) {
+    // Get all children names under path
+    List<String> children = _zkBaseDataAccessor.getChildNames(path, AccessOption.PERSISTENT);
+    if (children == null || children.isEmpty()) {
+      // The whole path has been deleted so return immediately
+      return;
+    }
+    filterChildrenNames(children, currentVersion);
+    List<String> pathsToDelete = getPathsToDelete(path, children);
+    for (String pathToDelete : pathsToDelete) {
+      // TODO: Should be batch delete but it doesn't work. It's okay since this runs async
+      _zkBaseDataAccessor.remove(pathToDelete, AccessOption.PERSISTENT);
+    }
   }
 
   /**
-   * Acquires the lock (create an ephemeral node) only if it is free (no ephemeral node already
-   * exists) at the time of invocation.
-   * @param path
+   * Filter out currentVersion and non-version children names from the children list given.
+   * @param children
+   * @param currentVersion
+   */
+  private void filterChildrenNames(List<String> children, long currentVersion) {
+    // Leave out metadata
+    children.remove(LAST_SUCCESSFUL_WRITE_KEY);
+    children.remove(LAST_WRITE_KEY);
+
+    // Keep the children that are not expired and exclude currentVersion from deletion
+    long currTime = System.currentTimeMillis();
+    children.removeIf(childName -> currTime < extractTimestamp(childName) + _versionTTL
```

Comment:
In case some other random nodes are added to the folder, will extractTimestamp(childName) and extractVersion(childName) fail with exception? I think we should remove those random nodes as well as a clean up. But that also means we need to validate if their names are valid or not before the filtering.

Reply:
On more comments, this usage of versionTTL may leave some garbage uncollected if no new clean up is triggered. Why not just sleep in the thread for versionTTL long and then clean up everything before the current version?

Reply:
Added validation. If invalid, it would just return Long.MAX_VALUE and add a warn log. Alternatively, I could have made it so that we return -1L so they will get cleaned up. Either way as long as things don't break.

Reply:
> On more comments, this usage of versionTTL may leave some garbage uncollected if no new clean up is triggered. Why not just sleep in the thread for versionTTL long and then clean up everything before the current version?

Can you reply to this one? I think scheduling the cleanup task for the future (after TTL) would be a cleaner way. It reduces the possibility of garbage leftovers. Moreover, you won't need to check (or maybe recording) the Timestamp in this way.

Reply:
@jiajunwang Sure. We could have a "delete scheduler" for cleanup.